Better error messages for unauthorized PATs to private packages

**Description**

<!--What needs to be improved? What should be done? Has there been any related work
before? Any relevant links or material?
-->

*disclaimer: im not sure if this is possible / makes sense considering how PATs work*

Trying to sign an image using a PAT gave me this error:
```
Error: signing [ghcr.io/ORG/IMAGE@sha256:DIGEST]: accessing image: entity not found in registry
main.go:74: error during command execution: signing [ghcr.io/ORG/IMAGE@sha256:DIGEST]: accessing image: entity not found in registry
```

I was sort of stumped as to why it happened, but it simply turned out I had not allowed the PAT access to my organization via the SAML login.

On a public package, this error was handled quite well, telling me that the PAT was not "logged in" using SAML. However, only a 404-esque response was given for private packages.

--- a/cmd/cosign/cli/sign/sign.go
+++ b/cmd/cosign/cli/sign/sign.go
@@ -179,7 +179,7 @@ func SignCmd(ro *options.RootOptions, ko options.KeyOpts, signOpts options.SignO
 
 		if digest, ok := ref.(name.Digest); ok && !signOpts.Recursive {
 			se, err := ociremote.SignedEntity(ref, opts...)
-			if err == ociremote.ErrEntityNotFound {
+			if ociremote.IsEntityNotFoundError(err) {
 				se = ociremote.SignedUnknown(digest)
 			} else if err != nil {
 				return fmt.Errorf("accessing image: %w", err)

--- a/cmd/cosign/cli/sign/sign.go
+++ b/cmd/cosign/cli/sign/sign.go
@@ -179,7 +179,7 @@ func SignCmd(ro *options.RootOptions, ko options.KeyOpts, signOpts options.SignO
 
 		if digest, ok := ref.(name.Digest); ok && !signOpts.Recursive {
 			se, err := ociremote.SignedEntity(ref, opts...)
-			if ociremote.IsEntityNotFoundError(err) {
+			if _, isEntityNotFoundErr := err.(*ociremote.EntityNotFoundError); isEntityNotFoundErr {
 				se = ociremote.SignedUnknown(digest)
 			} else if err != nil {
 				return fmt.Errorf("accessing image: %w", err)

--- a/pkg/oci/remote/remote.go
+++ b/pkg/oci/remote/remote.go
@@ -20,7 +20,6 @@ import (
 	"fmt"
 	"io"
 	"net/http"
-	"strings"
 
 	"github.com/google/go-containerregistry/pkg/name"
 	v1 "github.com/google/go-containerregistry/pkg/v1"
@@ -37,18 +36,22 @@ var (
 	remoteIndex = remote.Index
 	remoteGet   = remote.Get
 	remoteWrite = remote.Write
-
-	// ErrEntityNotFound is the error that SignedEntity returns when the
-	// provided ref does not exist.
-	ErrEntityNotFound = "cosign remoteGet: entity not found in registry"
 )
 
-func NewEntityNotFoundError(err error) error {
-	return fmt.Errorf("%s error: %w", ErrEntityNotFound, err)
+// EntityNotFoundError is the error that SignedEntity returns when the
+// provided ref does not exist.
+type EntityNotFoundError struct {
+	baseErr error
 }
 
-func IsEntityNotFoundError(err error) bool {
-	return strings.Contains(err.Error(), ErrEntityNotFound)
+func (e *EntityNotFoundError) Error() string {
+	return fmt.Sprintf("entity not found in registry, error: %v", e.baseErr)
+}
+
+func NewEntityNotFoundError(err error) error {
+	return &EntityNotFoundError{
+		baseErr: err,
+	}
 }
 
 // SignedEntity provides access to a remote reference, and its signatures.

--- a/pkg/oci/remote/remote.go
+++ b/pkg/oci/remote/remote.go
@@ -40,7 +40,7 @@ var (
 
 	// ErrEntityNotFound is the error that SignedEntity returns when the
 	// provided ref does not exist.
-	ErrEntityNotFound = "entity not found in registry"
+	ErrEntityNotFound = "cosign remoteGet: entity not found in registry"
 )
 
 func NewEntityNotFoundError(err error) error {

--- a/pkg/oci/remote/remote.go
+++ b/pkg/oci/remote/remote.go
@@ -44,7 +44,7 @@ var (
 )
 
 func NewEntityNotFoundError(err error) error {
-	return fmt.Errorf("%s error: %v", ErrEntityNotFound, err)
+	return fmt.Errorf("%s error: %w", ErrEntityNotFound, err)
 }
 
 func IsEntityNotFoundError(err error) bool {

--- a/pkg/oci/remote/remote.go
+++ b/pkg/oci/remote/remote.go
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"strings"
 
 	"github.com/google/go-containerregistry/pkg/name"
 	v1 "github.com/google/go-containerregistry/pkg/v1"
@@ -39,8 +40,16 @@ var (
 
 	// ErrEntityNotFound is the error that SignedEntity returns when the
 	// provided ref does not exist.
-	ErrEntityNotFound = errors.New("entity not found in registry")
+	ErrEntityNotFound = "entity not found in registry"
 )
+
+func NewEntityNotFoundError(err error) error {
+	return fmt.Errorf("%s error: %v", ErrEntityNotFound, err)
+}
+
+func IsEntityNotFoundError(err error) bool {
+	return strings.Contains(err.Error(), ErrEntityNotFound)
+}
 
 // SignedEntity provides access to a remote reference, and its signatures.
 // The SignedEntity will be one of SignedImage or SignedImageIndex.
@@ -50,7 +59,7 @@ func SignedEntity(ref name.Reference, options ...Option) (oci.SignedEntity, erro
 	got, err := remoteGet(ref, o.ROpt...)
 	var te *transport.Error
 	if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
-		return nil, ErrEntityNotFound
+		return nil, NewEntityNotFoundError(err)
 	} else if err != nil {
 		return nil, err
 	}

